---
title: "Gene Network Analysis using the SUFA Model"
output: rmarkdown::html_vignette
bibliography: ../inst/REFERENCES.bib
vignette: >
  %\VignetteIndexEntry{Gene Network Analysis using the SUFA Model}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We now turn to a case study on gene associations among immune cells.
Here, we integrate data from three studies analyzing gene expressions. The first is the GSE109125 bulkRNAseq dataset, collected from 103 highly purified immunocyte populations representing all lineages and several differentiation cascades and profiled using the ImmGen pipeline [@bulk_train]. The second study is a microarray dataset GSE15907 [@micro1_train1; @micro1_train2], measured on multiple *ex-vivo* immune lineages, primarily from adult B6 male mice. Finally, we include the GSE37448 [@micro2_train] microarray dataset, also part of the Immgen project.

## Load Data from the SUFA `R` package

```{r setup}
library(SUFA)
data(genedata)
```

An `R` list named `genedata` will be loaded into the environment. See the relevant documentation.

Then we set a seed for reproducibility:

```{r set_seed}
set.seed(1)
```

### Extract (`log2` transformed) gene expression datasets separately from `genedata`

```{r extraction}
bulk.data= genedata$bulk
array.data= genedata$array1
array.data2= genedata$array2
```

### Extract cell-types

```{r extraction_types}
Ident.bulk= genedata$bulk.types
Ident.array= genedata$array1.types
Ident.array2= genedata$array2.types
```

## Install and load the `genefilter R` package

```{r install_genefilter}
if (!require("genefilter", quietly = TRUE)){
  # if (!require("BiocManager", quietly = TRUE))
  #   install.packages("BiocManager")
  BiocManager::install("genefilter")
}
# BiocManager::install("genefilter")
library(Biobase);library(genefilter)
```

## Select top 5% genes from each dataset using `R::genefilter`

### BulkRNASeq data

```{r hvg_bulk}
dat=ExpressionSet(assayData = bulk.data)
filter.dat=varFilter(dat, var.cutoff = 0.95)
hvgs_in_bulk=rownames(filter.dat)
rm(filter.dat,dat)
```

### Microarray 1 data

```{r hvg_array1}
dat=ExpressionSet(assayData = array.data)
filter.dat=varFilter(dat, var.cutoff = 0.95)
hvgs_in_array=rownames(filter.dat)
rm(filter.dat,dat)
```

### Microarray 2 data

```{r hvg_array2}
dat=ExpressionSet(assayData = array.data2)
filter.dat=varFilter(dat, var.cutoff = 0.95)
hvgs_in_array2=rownames(filter.dat)
rm(filter.dat,dat)
```

Consider the intersection of the filtered genes across the studies for the analysis

```{r}
genes.use=Reduce(intersect,list(hvgs_in_array,hvgs_in_array2,hvgs_in_bulk))
length(genes.use)
```

We have 474 genes.

## Heatmap of the datasets

Since different cell-types exhibit very different gene expression profiles, we centered the gene-expressions separately within each cell type.

```{r gene_heatmap, fig.dim = c(7, 5), fig.cap="Heatmaps of two microarray and a bulk RNAseq dataset on a common set of genes. The X and Y axes represent genes and cells across studies, respectively."}
Y=list(t(array.data[genes.use,]), t(array.data2[genes.use,]),t(bulk.data[genes.use,] ))
Ident.list=list(Ident.array,Ident.array2,Ident.bulk)

Y_cent=mapply(function(Idents,y){
  cell.types=unique(Idents)
  y_cent=lapply(cell.types, function(ident,Identt,Y){
    scale(Y[which(Identt==ident),],scale=F)
  } ,Y=y,Identt=Idents)
  Reduce(rbind,y_cent)
}, Ident.list, Y )

Y_ordered=mapply( function(y, name){
  d=dist(y)
  hc2=cluster::agnes(d)
  library(reshape2)
  dimnames(y)=NULL
  melted_cormat <- melt(t((y[hc2$order,]/max(abs(y)))))
  melted_cormat$Dataset=as.factor(name)
  melted_cormat
},Y_cent,c("Microarray 1", "Microarray 2", "Bulk RNAseq"),SIMPLIFY = F)

Y_ordered=Reduce(rbind,Y_ordered)
library(ggplot2)
ggplot(data = Y_ordered, aes(x=Var1, y=Var2, fill=value))+  geom_tile(aes(fill = value)) +
  theme (panel.grid.major = element_blank(), panel.border = element_blank(),
         panel.background = element_blank(), legend.position = "none",text = element_text(size=15))+ 
  scale_fill_viridis_c(option="B",alpha=1) +  facet_grid(rows=vars(Dataset),scales = "free")+
  xlab("Genes")+ylab("Cells")
```

```{r echo=FALSE}
rm(Y_ordered)
```

## Fit SUFA on the datasets

### Pre-processing data

We scale the data with respect to the median of the sample standard deviations of the genes across all studies.

```{r scaling}
sd.median=median( sapply(Y_cent, function(y) apply(y,2,sd)))
Y_cent_scaled=lapply(Y_cent, "/",   sd.median)
```

### Fit SUFA

```{r fit_SUFA, message=FALSE, results=FALSE, warning=FALSE, comment=FALSE}
res.sufa<-fit_SUFA(Y_cent_scaled,qmax=25,nthreads = 5,nrun = 7.5e3,
                                     nleapfrog = 4, leapmax = 9, del_range = c(0,.009))
```

We discard the first 2,500 iterations as the burnin samples. Since we used a thinning interval of 5 to store the MCMC samples, we set `burnin=2500/5=500`.

```{r set_burnin}
burn=500
```

Obtain the widely application Bayesian information criterion (WBIC) [@wbic2013]:

```{r wbic}
WBIC(res.sufa ,Y_cent_scaled, model="SUFA",burn=5e2,ncores=1)
```

## Obtain the shared correlation matrix

We use the uncertainty quantified by the posterior samples to infer whether correlations are zero---we encode the off-diagonals of the correlation matrix as zero if the respective 95% posterior credible interval contains zero. This leads to a distinctive sparsity pattern.

```{r get_cormat}
phis=apply(res.sufa$Lambda[,,-(1:burn)],3,identity, simplify = F)
diags=apply(res.sufa$residuals[-(1:burn),],1,identity, simplify = F)

cormats=abind::abind( mapply(function(phi, ps) cov2cor( tcrossprod(phi)+diag(ps)) 
                             ,phis,diags, SIMPLIFY = F),along = 3 )
cormat_mean=apply(cormats, c(1, 2), mean)

alpha=.05
low = apply(cormats, c(1, 2), quantile, alpha/2)
high = apply(cormats, c(1, 2), quantile, 1 - alpha/2)
cormat_mean[(low < 0) & (high > 0)]=0
```

```{r plot_cormat ,echo=FALSE,fig.dim= c(5.5,4.5), fig.cap ="Shared correlation structre of the genes"}
C2=cormat_mean; 
d <- as.dist(1-abs(C2)) 
# Hierarchical clustering using Complete Linkage
hc2=cluster::agnes(d)
C2[cormat_mean==0]=NA
heatplot(C2[hc2$order,hc2$order])

rm(C2)
```

## Identifying important hub genes

Next, we focus on identifying important hub genes that have absolute correlation $\geq 0.25$ with at least 10 other genes.

```{r subgenes}
Z2=apply(cormats, c(1, 2), mean)
rownames(Z2)=colnames(Z2)=genes.use
Z2[abs(Z2)<.25]=0
indices.connected.genes=which(rowSums( Z2!=0)-1 >10 )
subgenes=genes.use[indices.connected.genes]
```

The resulting dependency network is summarized as a circos plot [@gu2014circlize] in the following figure:

```{r circos_plot, echo=FALSE, message=FALSE, warning=FALSE, comment=FALSE, dpi=200,out.width="120%", fig.cap ="Circos plot of the dependency structure between genes, blue (red) connection implies a positive (negative), their opacities being proportional to the corresponding association strengths. Associations are only plotted for absolute correlation >= 0.25."}
library(circlize)
C2=Z2[subgenes,subgenes]; 
d <- as.dist(1-abs(C2)) 
# Hierarchical clustering using Complete Linkage
hc2=cluster::agnes(d)
indices.connected.genes.ordered=hc2$order
subgenes.connected.ordered=subgenes[indices.connected.genes.ordered]

est_partcor_plot_subgenes_connected_ordered=C2[subgenes.connected.ordered,subgenes.connected.ordered] 
col_red_blue = colorRamp2(c(-.5,0,.5), c("red","white", "blue"), transparency = 0.00)
diag(est_partcor_plot_subgenes_connected_ordered)=.0001
par(cex=.2)
chordDiagramFromMatrix(est_partcor_plot_subgenes_connected_ordered, order=c(subgenes.connected.ordered), symmetric=TRUE, transparency = 0.25, grid.col="blue", col=col_red_blue, annotationTrack=c("grid"), keep.diagonal=T, scale=T, reduce=-100, preAllocateTracks = list(track.height = max(strwidth(unlist(dimnames(est_partcor_plot_subgenes_connected_ordered))))))
circos.track(track.index = 1, panel.fun = function(x, y) {
  circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
              facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
}, bg.border = NA)
```

## Shared and study-specific factor loading matrices

We visualize the six columns having largest sum-of-squares from the shared loading matrix, and repeat for the top three columns of each study-specific loading matrix.

```{r loadings, message=FALSE, warning=FALSE, comment=FALSE,out.width="120%", fig.height=7, dpi=300, fig.cap ="Plots the top six columns with highest sum-of-squares values from the shared loading matrix, and repeat for the top three columns of each study-specific loading matrix."}
library(dplyr);library(ggplot2);library(dplyr);library(forcats);library(viridis)
# studies=c("Microarray 1", "Microarray 2","BulkRNASeq")
studies=c("Array 1", "Array 2","Bulk")
loadings=lam.est.all(res.sufa,burn = burn)

loadings_shared=data.frame(Loads= loadings$Shared[,1:6],Genes=genes.use, Loading="Shared");
loadings_study=foreach(s =1:length(loadings$Study_specific),.combine = rbind,.packages = c("SUFA","abind")) %do%{
  data.frame(Loads= loadings$Study_specific[[s]][,1:3] ,Genes=genes.use, Loading=studies[s])
}
loadings.m=rbind(melt(loadings_shared,id.vars = c("Loading","Genes")),
melt(loadings_study,id.vars = c("Loading","Genes")) )
loadings.m=subset(loadings.m, subset=Genes%in% (subgenes.connected.ordered<-subgenes[hc2$order]) )
loadings.m%>%
  mutate(Genes = fct_relevel(Genes,
                             subgenes[hc2$order] ),
         Loading= fct_relevel(Loading,c("Shared","Bulk","Array 1","Aarray 2") ) ) %>%
  ggplot( aes(Genes, (value), fill=value)) + 
  facet_wrap(~Loading+ variable, nrow=1,scales="free_x") + #place the factors in separate facets
  geom_bar(stat="identity") + ylab("Loading Strength") +coord_flip() + 
  theme_bw(base_size=10)+
  theme(legend.position = "bottom",axis.text.y = element_text(hjust = 1,size=2),
        axis.text.x = element_text(size=2, angle = 90), strip.text.x = element_text(size = 2),
        legend.key.height=unit(.01,"in"),legend.key.width=unit(.3,"in"), 
        legend.title = element_text(size = 2.5), legend.text = element_text(size = 2.5), 
        axis.title.x = element_text(size = 3), axis.title.y = element_text(size = 3))+
  scale_fill_viridis("Loading", option = "A") ->pplot
pplot
```

## References
